zeppelin-web-angular/e2e/models/notebook-paragraph-page.util.ts

```diff
@@ -0,0 +1,182 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { expect, Page, test } from '@playwright/test';
+import { NotebookParagraphPage } from './notebook-paragraph-page';
+
+export class NotebookParagraphUtil {
+  private page: Page;
+  private paragraphPage: NotebookParagraphPage;
+
+  constructor(page: Page) {
+    this.page = page;
+    this.paragraphPage = new NotebookParagraphPage(page);
+  }
+
+  async verifyParagraphContainerStructure(): Promise<void> {
+    await expect(this.paragraphPage.paragraphContainer).toBeVisible();
+    await expect(this.paragraphPage.controlPanel).toBeVisible();
+  }
+
+  async verifyDoubleClickEditingFunctionality(): Promise<void> {
+    await expect(this.paragraphPage.paragraphContainer).toBeVisible();
+
+    await this.paragraphPage.doubleClickToEdit();
+
+    await expect(this.paragraphPage.codeEditor).toBeVisible();
+  }
+
+  async verifyAddParagraphButtons(): Promise<void> {
+    // Verify "Add Paragraph Above" button
+    await expect(this.paragraphPage.addParagraphAbove).toBeVisible();
+    const addAboveCount = await this.paragraphPage.addParagraphAbove.count();
+    expect(addAboveCount).toBe(1); // Expect exactly one "Add Above" button
+    console.log(`✓ Add Paragraph Above button verified (count: ${addAboveCount})`);
+
+    // Verify "Add Paragraph Below" button
+    await expect(this.paragraphPage.addParagraphBelow).toBeVisible();
+    const addBelowCount = await this.paragraphPage.addParagraphBelow.count();
+    expect(addBelowCount).toBe(1); // Expect exactly one "Add Below" button
+    console.log(`✓ Add Paragraph Below button verified (count: ${addBelowCount})`);
+  }
+
+  async verifyParagraphControlInterface(): Promise<void> {
+    await expect(this.paragraphPage.controlPanel).toBeVisible();
+    await this.paragraphPage.runButton.isVisible();
+    await expect(this.paragraphPage.runButton).toBeVisible();
+    const isRunEnabled = await this.paragraphPage.isRunButtonEnabled();
+    expect(isRunEnabled).toBe(true);
+  }
+
+  async verifyCodeEditorFunctionality(): Promise<void> {
+    await this.paragraphPage.isCodeEditorVisible();
+    await expect(this.paragraphPage.codeEditor).toBeVisible();
+  }
+
+  async verifyResultDisplaySystem(): Promise<void> {
+    await this.paragraphPage.hasResult();
+    await expect(this.paragraphPage.resultDisplay).toBeVisible();
+  }
```

Comment:
Not testing the result of the expression `await this paragraphPage.hasResult()` has no effect to the method result.